File: modules/expo-pinned-websocket/android/src/main/java/expo/modules/pinnedwebsocket/ExpoPinnedWebsocketModule.kt
package expo.modules.pinnedwebsocket

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.*
import okio.ByteString
import java.security.MessageDigest
import java.security.cert.X509Certificate
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import javax.net.ssl.*

class ExpoPinnedWebsocketModule : Module() {

    private val activeSockets = ConcurrentHashMap<Int, PinnedWebSocketImpl>()

    override fun definition(): ModuleDefinition = ModuleDefinition {
        Name("ExpoPinnedWebsocket")

        Events("onOpen", "onMessage", "onClose", "onError", "onPeerSpki", "onPinError")

        Function("createSocket") { socketId: Int, urlString: String, allowedSpkiHashes: List<String> ->
            val impl = PinnedWebSocketImpl(
                socketId = socketId,
                url = urlString,
                allowedSpkiHashes = allowedSpkiHashes,
                onEvent = { name, payload -> sendEvent(name, payload) }
            )
            activeSockets[socketId] = impl
            impl.connect()
        }

        Function("sendMessage") { socketId: Int, data: String ->
            activeSockets[socketId]?.send(data)
        }

        Function("closeSocket") { socketId: Int ->
            activeSockets.remove(socketId)?.close()
        }
    }
}

// ── PinnedWebSocketImpl ──────────────────────────────────────────────────────

internal class PinnedWebSocketImpl(
    private val socketId: Int,
    private val url: String,
    private val allowedSpkiHashes: List<String>,
    private val onEvent: (String, Map<String, Any?>) -> Unit
) {
    private var webSocket: WebSocket? = null
    // AtomicBoolean ensures JMM visibility and provides compareAndSet for exactly-once
    // semantics — onClosed / onFailure / close() may fire concurrently from OkHttp's
    // thread pool and the Expo module function queue.
    private val closed = java.util.concurrent.atomic.AtomicBoolean(false)

    fun connect() {
        val scheme = try { java.net.URI(url).scheme?.lowercase() } catch (_: Exception) { null }
        require(scheme == "ws" || scheme == "wss") {
            "expo-pinned-websocket: URL scheme must be ws:// or wss://"
        }

        val (trustManager, sslSocketFactory) = buildTlsComponents()

        val client = OkHttpClient.Builder()
            .sslSocketFactory(sslSocketFactory, trustManager)
            // Hostname verification uses OkHttp's default — no override needed.
            // OkHostnameVerifier (internal API) is intentionally omitted; OkHttp
            // applies its default hostname verifier when sslSocketFactory is set.
            // Transport-level keepalive: OkHttp sends a WebSocket ping every 30s and
            // fails the socket if a pong isn't received within the same interval.
            // This detects half-open sockets (Wi-Fi hand-off, NAT timeout) faster than
            // the 30s app-level tick watchdog alone.
            .pingInterval(30, TimeUnit.SECONDS)
            .build()

        // OkHttp does not send an Origin header by default. The OpenClaw gateway
        // enforces controlUi.allowedOrigins, so we mirror what browsers send:
        // the HTTP(S) form of the WebSocket URL's host[:port] (RFC 6454).
        val request = Request.Builder()
            .url(url)
            .header("Origin", httpOrigin(url))
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                onEvent("onOpen", mapOf("socketId" to socketId))
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                onEvent("onMessage", mapOf("socketId" to socketId, "data" to text))
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                val text = bytes.utf8()
                if (BuildConfig.DEBUG && text.toByteArray(Charsets.UTF_8).size != bytes.size()) {
                    android.util.Log.w(
                        "PinnedWebSocket",
                        "Binary frame (socketId=$socketId) contained non-UTF-8 bytes — " +
                            "invalid sequences were replaced with \uFFFD"
                    )
                }
                onEvent("onMessage", mapOf("socketId" to socketId, "data" to text))
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                if (closed.compareAndSet(false, true)) {
                    onEvent("onClose", mapOf(
                        "socketId" to socketId,
                        "code" to code,
                        "reason" to reason,
                        "wasClean" to true
                    ))
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (closed.compareAndSet(false, true)) {
                    onEvent("onError", mapOf("socketId" to socketId, "message" to (t.message ?: "Unknown error")))
                    onEvent("onClose", mapOf(
                        "socketId" to socketId,
                        "code" to 1006,
                        "reason" to (t.message ?: ""),
                        "wasClean" to false
                    ))
                }
            }
        })
    }

    fun send(data: String) {
        webSocket?.send(data)
    }

    fun close() {
        if (closed.compareAndSet(false, true)) {
            webSocket?.close(1000, "Normal closure")
        }
    }

    // ── TLS: SPKI extraction + pinning ─────────────────────────────────────

    private fun buildTlsComponents(): Pair<X509TrustManager, SSLSocketFactory> {
        val defaultTrustManager = getDefaultTrustManager()

        val pinnedTrustManager = object : X509TrustManager {
            override fun getAcceptedIssuers(): Array<X509Certificate> =
                defaultTrustManager.getAcceptedIssuers()

            override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) =
                defaultTrustManager.checkClientTrusted(chain, authType)

            override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
                // Standard validation first — throws if chain is not trusted by the OS.
                defaultTrustManager.checkServerTrusted(chain, authType)

                // Extract SPKI SHA-256 of the leaf cert.
                val leaf = chain[0]
                val spkiHash = spkiSha256Hex(leaf)

                if (spkiHash == null) {
                    // Cannot extract SPKI — fail closed when pins are configured.
                    if (allowedSpkiHashes.isNotEmpty()) {
                        throw javax.net.ssl.SSLPeerUnverifiedException(
                            "Unable to extract SPKI from leaf cert for pin enforcement"
                        )
                    }
                    // TOFU mode: cannot observe, let through silently.
                    return
                }

                onEvent("onPeerSpki", mapOf("socketId" to socketId, "sha256Hex" to spkiHash))

                if (allowedSpkiHashes.isEmpty()) return // TOFU mode: observe only

                if (!allowedSpkiHashes.any { constantTimeEquals(it, spkiHash) }) {
                    onEvent("onPinError", mapOf(
                        "socketId" to socketId,
                        "observed" to spkiHash,
                        "allowed" to allowedSpkiHashes
                    ))
                    throw javax.net.ssl.SSLPeerUnverifiedException(
                        "Certificate pin mismatch. Observed: $spkiHash"
                    )
                }
            }
        }

        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, arrayOf(pinnedTrustManager), null)
        return Pair(pinnedTrustManager, sslContext.socketFactory)
    }

    companion object {
        /**
         * Compares two strings in constant time using [MessageDigest.isEqual] to
         * prevent timing side-channels during SPKI pin verification. Network jitter
         * makes this practically unexploitable for SPKI pins, but the audit checklist
         * explicitly requires constant-time comparison.
         */
        fun constantTimeEquals(a: String, b: String): Boolean {
            val aBytes = a.toByteArray(Charsets.UTF_8)
            val bBytes = b.toByteArray(Charsets.UTF_8)
            return MessageDigest.isEqual(aBytes, bBytes)
        }

        /**
         * Returns the HTTP(S) origin string for a WebSocket URL, matching the
         * behaviour of browsers and SocketRocket (RN's built-in iOS WebSocket):
         * scheme mapped to http/https, host preserved, port included only when
         * it differs from the scheme default (443 for wss, 80 for ws).
         */
        fun httpOrigin(urlString: String): String {
            return try {
                val parsed = java.net.URI(urlString)
                val scheme = when (parsed.scheme?.lowercase()) {
                    "wss", "https" -> "https"
                    else -> "http"
                }
                val defaultPort = if (scheme == "https") 443 else 80
                val port = parsed.port
                if (port > 0 && port != defaultPort) {
                    "$scheme://${parsed.host}:$port"
                } else {
                    "$scheme://${parsed.host}"
                }
            } catch (_: Exception) {
                // Fallback: the URL was already validated in the Expo module Function.
                urlString
            }
        }

        private fun getDefaultTrustManager(): X509TrustManager {
            val factory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
            factory.init(null as java.security.KeyStore?)
            return factory.trustManagers.filterIsInstance<X509TrustManager>().first()
        }

        /**
         * Returns the lowercase hex SHA-256 of the SubjectPublicKeyInfo (SPKI)
         * DER encoding of [cert]'s public key.
         *
         * `cert.publicKey.encoded` on Android returns the full SPKI DER
         * (X.509 SubjectPublicKeyInfo encoding) — this is the correct input for
         * SPKI pin comparison and matches Chrome/Firefox/iOS expectations.
         */
        fun spkiSha256Hex(cert: X509Certificate): String? {
            return try {
                val spkiDer = cert.publicKey.encoded ?: return null
                val hash = MessageDigest.getInstance("SHA-256").digest(spkiDer)
                hash.joinToString("") { "%02x".format(it) }
            } catch (_: java.security.NoSuchAlgorithmException) {
                // SHA-256 is mandatory on Android (java.security spec §1.4.2); should never happen.
                null
            }
        }
    }
}

File: modules/expo-pinned-websocket/ios/ExpoPinnedWebsocketModule.swift
import ExpoModulesCore
import CryptoKit

// ── ASN.1 DER helpers ────────────────────────────────────────────────────────
//
// Minimal parser used only to locate SubjectPublicKeyInfo inside a certificate
// DER blob. We avoid a full ASN.1 library to keep the module self-contained.

/// Decodes one DER element starting at `offset` in `der`.
/// Returns (elementStart, contentStart, end) or nil if the encoding is invalid.
private func derElement(_ der: Data, at offset: Int) -> (elementStart: Int, contentStart: Int, end: Int)? {
  guard offset < der.count else { return nil }
  var i = offset + 1 // skip tag byte
  guard i < der.count else { return nil }
  let firstLen = Int(der[i])
  i += 1
  let length: Int
  if firstLen & 0x80 == 0 {
    length = firstLen
  } else {
    let n = firstLen & 0x7f
    guard n > 0, i + n <= der.count else { return nil }
    var len = 0
    for _ in 0..<n { len = (len << 8) | Int(der[i]); i += 1 }
    length = len
  }
  let end = i + length
  guard end <= der.count else { return nil }
  return (offset, i, end)
}

/// Walks the RFC 5280 ASN.1 certificate DER and returns the raw bytes of the
/// SubjectPublicKeyInfo SEQUENCE, or nil if the structure cannot be parsed.
///
/// TBSCertificate field order (RFC 5280 §4.1):
///   [0] version (optional), serialNumber, signature, issuer,
///   validity, subject, subjectPublicKeyInfo, ...
private func extractSpkiDer(from certDer: Data) -> Data? {
  // Outer Certificate SEQUENCE
  guard let cert = derElement(certDer, at: 0), certDer[cert.elementStart] == 0x30 else { return nil }
  // TBSCertificate SEQUENCE
  guard let tbs = derElement(certDer, at: cert.contentStart), certDer[tbs.elementStart] == 0x30 else { return nil }

  var pos = tbs.contentStart

  // Skip optional version [0] EXPLICIT (context-constructed tag 0xa0)
  if pos < certDer.count && certDer[pos] == 0xa0 {
    guard let e = derElement(certDer, at: pos) else { return nil }
    pos = e.end
  }
  // serialNumber INTEGER (0x02)
  guard pos < certDer.count, certDer[pos] == 0x02,
        let sn = derElement(certDer, at: pos) else { return nil }
  pos = sn.end
  // signature AlgorithmIdentifier SEQUENCE (0x30)
  guard pos < certDer.count, certDer[pos] == 0x30,
        let sig = derElement(certDer, at: pos) else { return nil }
  pos = sig.end
  // issuer Name SEQUENCE (0x30)
  guard pos < certDer.count, certDer[pos] == 0x30,
        let issuer = derElement(certDer, at: pos) else { return nil }
  pos = issuer.end
  // validity SEQUENCE (0x30)
  guard pos < certDer.count, certDer[pos] == 0x30,
        let validity = derElement(certDer, at: pos) else { return nil }
  pos = validity.end
  // subject Name SEQUENCE (0x30)
  guard pos < certDer.count, certDer[pos] == 0x30,
        let subject = derElement(certDer, at: pos) else { return nil }
  pos = subject.end
  // subjectPublicKeyInfo SEQUENCE — this is what we want
  guard pos < certDer.count, certDer[pos] == 0x30,
        let spki = derElement(certDer, at: pos) else { return nil }

  return certDer[spki.elementStart..<spki.end]
}

/// Returns the lowercase hex SHA-256 of the SubjectPublicKeyInfo DER of the
/// leaf certificate in `serverTrust`, or nil if extraction fails.
///
/// Parses the SPKI directly from the certificate DER (via SecCertificateCopyData)
/// rather than reconstructing it from a key-type header table. This is
/// key-type-agnostic and produces the same hash as Chrome, Firefox, and Android
/// for all key types: RSA (any size), EC (any curve), Ed25519, etc.
private func spkiSha256Hex(from serverTrust: SecTrust) -> String? {
  guard let chain = SecTrustCopyCertificateChain(serverTrust) as? [SecCertificate],
        let cert = chain.first else { return nil }
  let certDer = SecCertificateCopyData(cert) as Data
  guard let spkiDer = extractSpkiDer(from: certDer) else { return nil }
  let digest = SHA256.hash(data: spkiDer)
  return digest.map { String(format: "%02x", $0) }.joined()
}

// ── Origin helper ───────────────────────────────────────────────────────────

/// Returns an HTTP(S) origin string for a `ws://`/`wss://` URL: scheme-mapped
/// to `http`/`https`, host preserved, and port included only when it differs
/// from the scheme's default (443 for https, 80 for http).
///
/// This matches SocketRocket's `RCTSR_origin` implementation so that standard
/// gateway origin-allow-list entries (`https://example.com`) work for both the
/// built-in RN WebSocket and this module regardless of whether the user omits
/// the port in the URL.
private func httpOrigin(for url: URL) -> String? {
  guard let scheme = url.scheme?.lowercased(), let host = url.host else { return nil }
  let httpScheme: String
  let defaultPort: Int
  switch scheme {
  case "wss", "https": httpScheme = "https"; defaultPort = 443
  case "ws",  "http":  httpScheme = "http";  defaultPort = 80
  default:             return nil
  }
  if let port = url.port, port != defaultPort {
    return "\(httpScheme)://\(host):\(port)"
  }
  return "\(httpScheme)://\(host)"
}

// ── Constant-time comparison ─────────────────────────────────────────────────

/// Compares two strings in constant time (XOR all byte pairs regardless of
/// early mismatch) to prevent timing side-channels during SPKI pin verification.
/// Network jitter makes this practically unexploitable for SPKI pins, but the
/// audit checklist explicitly requires constant-time comparison.
private func secureEquals(_ a: String, _ b: String) -> Bool {
  let aBytes = Array(a.utf8)
  let bBytes = Array(b.utf8)
  guard aBytes.count == bBytes.count else { return false }
  return zip(aBytes, bBytes).reduce(UInt8(0)) { $0 | ($1.0 ^ $1.1) } == 0
}

// ── PinnedSocketDelegate ────────────────────────────────────────────────────

/// Manages one `URLSessionWebSocketTask` and bridges its lifecycle events back
/// to the Expo module for forwarding to JS.
class PinnedSocketDelegate: NSObject, URLSessionDelegate, URLSessionWebSocketDelegate, URLSessionTaskDelegate {

  private let socketId: Int
  private let allowedSpkiHashes: [String]
  private let onEvent: (String, [String: Any]) -> Void

  private var task: URLSessionWebSocketTask?
  private var session: URLSession?
  private let _closedLock = NSLock()
  private var _closed = false
  /// NSLock-guarded accessor — `closed` is written from three concurrent contexts:
  /// the URLSession delegate queue, the main thread (ping timer), and the Expo
  /// module function queue. All read/write access goes through this computed property.
  private var closed: Bool {
    get { _closedLock.withLock { _closed } }
    set { _closedLock.withLock { _closed = newValue } }
  }
  private var pingTimer: Timer?
  /// Transport-level ping interval, matching OkHttp's default on Android.
  private static let pingIntervalSeconds: TimeInterval = 30

  init(
    socketId: Int,
    url: URL,
    allowedSpkiHashes: [String],
    onEvent: @escaping (String, [String: Any]) -> Void
  ) {
    self.socketId = socketId
    self.allowedSpkiHashes = allowedSpkiHashes
    self.onEvent = onEvent
    super.init()

    let config = URLSessionConfiguration.default
    config.timeoutIntervalForRequest = 30
    // Use a private serial queue instead of .main so that TLS challenges,
    // message receives, and close events don't block the UI thread.
    // This mirrors SocketRocket's design (private queue + main-thread callbacks).
    let delegateQueue = OperationQueue()
    delegateQueue.qualityOfService = .userInitiated
    delegateQueue.maxConcurrentOperationCount = 1
    let session = URLSession(configuration: config, delegate: self, delegateQueue: delegateQueue)
    self.session = session

    // OpenClaw gateway enforces controlUi.allowedOrigins. URLSessionWebSocketTask
    // doesn't auto-populate Origin (unlike browsers / SocketRocket), so we mirror
    // the behavior browsers use for same-origin connections by sending the HTTP(S)
    // form of the WebSocket URL's host[:port].
    var request = URLRequest(url: url)
    if let origin = httpOrigin(for: url) {
      request.setValue(origin, forHTTPHeaderField: "Origin")
    }

    let task = session.webSocketTask(with: request)
    self.task = task
    task.resume()
  }

  // URLSessionDelegate — TLS challenge
  func urlSession(
    _ session: URLSession,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
          let serverTrust = challenge.protectionSpace.serverTrust
    else {
      completionHandler(.performDefaultHandling, nil)
      return
    }

    // Standard OS validation first.
    var error: CFError?
    let trusted = SecTrustEvaluateWithError(serverTrust, &error)
    guard trusted else {
      completionHandler(.cancelAuthenticationChallenge, nil)
      return
    }

    guard let observedHash = spkiSha256Hex(from: serverTrust) else {
      if allowedSpkiHashes.isEmpty {
        // TOFU mode: cannot compute SPKI hash, allow but skip recording the observation.
        completionHandler(.useCredential, URLCredential(trust: serverTrust))
      } else {
        // Pinning active: cannot verify — fail closed.
        completionHandler(.cancelAuthenticationChallenge, nil)
      }
      return
    }

    onEvent("onPeerSpki", ["socketId": socketId, "sha256Hex": observedHash])

    if allowedSpkiHashes.isEmpty {
      // TOFU mode: no enforcement, just observe.
      completionHandler(.useCredential, URLCredential(trust: serverTrust))
      return
    }

    if allowedSpkiHashes.contains(where: { secureEquals($0, observedHash) }) {
      completionHandler(.useCredential, URLCredential(trust: serverTrust))
    } else {
      onEvent("onPinError", [
        "socketId": socketId,
        "observed": observedHash,
        "allowed": allowedSpkiHashes
      ])
      completionHandler(.cancelAuthenticationChallenge, nil)
    }
  }

  // URLSessionWebSocketDelegate — open
  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    guard !closed else { return }
    onEvent("onOpen", ["socketId": socketId])
    startPingTimer()
    listen()
  }

  // URLSessionWebSocketDelegate — close
  func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    guard !closed else { return }
    closed = true
    stopPingTimer()
    let reasonStr = reason.flatMap { String(data: $0, encoding: .utf8) } ?? ""
    onEvent("onClose", [
      "socketId": socketId,
      "code": closeCode.rawValue,
      "reason": reasonStr,
      "wasClean": closeCode == .normalClosure || closeCode == .goingAway
    ])
    self.session?.invalidateAndCancel()
  }

  // URLSessionTaskDelegate — completion (connection error path)
  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    guard !closed else { return }
    closed = true
    stopPingTimer()
    if let error = error {
      // When the server rejects the WebSocket upgrade with an HTTP error (e.g. 403
      // "origin not allowed", 401 "unauthorized"), the task's response holds the
      // real status code. Surface it so the JS error handler can distinguish a
      // transport failure (1006) from an explicit server rejection (4000+status).
      let httpStatus = (task.response as? HTTPURLResponse).map { $0.statusCode }
      let closeCode: Int
      let reason: String
      if let status = httpStatus, status >= 400 {
        // Use the private close-code range 4000–4999 to encode HTTP status codes
        // from failed handshakes. The JS layer (OpenClawClient) treats any non-1000/1001
        // close as an error and will build a meaningful message from the reason string.
        closeCode = 4000 + status
        reason = "HTTP \(status): \(error.localizedDescription)"
      } else {
        closeCode = 1006
        reason = error.localizedDescription
      }
      onEvent("onError", ["socketId": socketId, "message": reason])
      onEvent("onClose", ["socketId": socketId, "code": closeCode, "reason": reason, "wasClean": false])
    }
    self.session?.invalidateAndCancel()
  }

  func send(_ data: String) {
    guard !closed else { return }
    task?.send(.string(data)) { [weak self] error in
      if let error, let self, !self.closed {
        self.onEvent("onError", ["socketId": self.socketId, "message": error.localizedDescription])
      }
    }
  }

  func close() {
    guard !closed else { return }
    closed = true
    stopPingTimer()
    task?.cancel(with: .normalClosure, reason: nil)
    session?.invalidateAndCancel()
  }

  // ── Transport-level ping ────────────────────────────────────────────────────

  private func startPingTimer() {
    stopPingTimer()
    // Timer must be scheduled on the main run loop regardless of the delegate
    // queue so it fires reliably even when there is no active network activity.
    DispatchQueue.main.async { [weak self] in
      guard let self, !self.closed else { return }
      let timer = Timer.scheduledTimer(
        withTimeInterval: PinnedSocketDelegate.pingIntervalSeconds,
        repeats: true
      ) { [weak self] _ in
        self?.sendPing()
      }
      self.pingTimer = timer
    }
  }

  private func stopPingTimer() {
    // Timer was scheduled on the main run loop; it must be invalidated there.
    let timer = pingTimer
    pingTimer = nil
    DispatchQueue.main.async { timer?.invalidate() }
  }

  private func sendPing() {
    guard !closed else { stopPingTimer(); return }
    task?.sendPing { [weak self] error in
      guard let self else { return }
      if let error, !self.closed {
        self.closed = true
        self.stopPingTimer()
        self.onEvent("onError", ["socketId": self.socketId, "message": "Ping failed: \(error.localizedDescription)"])
        self.onEvent("onClose", [
          "socketId": self.socketId,
          "code": 1006,
          "reason": "Ping failed: \(error.localizedDescription)",
          "wasClean": false
        ])
        self.session?.invalidateAndCancel()
      }
    }
  }

  // Read messages in a loop using the completion-handler API.
  private func listen() {
    guard !closed else { return }
    task?.receive { [weak self] result in
      guard let self, !self.closed else { return }
      switch result {
      case .success(let message):
        switch message {
        case .string(let text):
          self.onEvent("onMessage", ["socketId": self.socketId, "data": text])
        case .data(let data):
          if let text = String(data: data, encoding: .utf8) {
            self.onEvent("onMessage", ["socketId": self.socketId, "data": text])
          } else {
            #if DEBUG
            print("[PinnedWebSocket] Binary frame on socketId=\(self.socketId) could not be decoded as UTF-8 — dropped")
            #endif
          }
        @unknown default:
          break
        }
        self.listen()
      case .failure(let error):
        if !self.closed {
          self.closed = true
          self.stopPingTimer()
          self.onEvent("onError", ["socketId": self.socketId, "message": error.localizedDescription])
          self.onEvent("onClose", [
            "socketId": self.socketId,
            "code": 1006,
            "reason": error.localizedDescription,
            "wasClean": false
          ])
          self.session?.invalidateAndCancel()
        }
      }
    }
  }
}

// ── Expo Module ────────────────────────────────────────────────────────────

public class ExpoPinnedWebsocketModule: Module {
  private var sockets: [Int: PinnedSocketDelegate] = [:]

  public func definition() -> ModuleDefinition {
    Name("ExpoPinnedWebsocket")

    Events("onOpen", "onMessage", "onClose", "onError", "onPeerSpki", "onPinError")

    Function("createSocket") { [weak self] (socketId: Int, urlString: String, allowedSpkiHashes: [String]) in
      guard let self else { return }
      guard let url = URL(string: urlString) else {
        self.sendEvent("onError", [
          "socketId": socketId,
          "message": "Invalid URL: \(urlString)"
        ])
        return
      }
      guard let scheme = url.scheme?.lowercased(), scheme == "ws" || scheme == "wss" else {
        self.sendEvent("onError", [
          "socketId": socketId,
          "message": "expo-pinned-websocket: URL scheme must be ws:// or wss://"
        ])
        return
      }

      let delegate = PinnedSocketDelegate(
        socketId: socketId,
        url: url,
        allowedSpkiHashes: allowedSpkiHashes
      ) { [weak self] eventName, payload in
        self?.sendEvent(eventName, payload)
      }
      self.sockets[socketId] = delegate
    }

    Function("sendMessage") { [weak self] (socketId: Int, data: String) in
      self?.sockets[socketId]?.send(data)
    }

    Function("closeSocket") { [weak self] (socketId: Int) in
      self?.sockets[socketId]?.close()
      self?.sockets.removeValue(forKey: socketId)
    }
  }
}
